### PR TITLE
Fix issue 401

### LINKED ISSUE #401: Validate interpreter handling of stack unwinds
While debugging `fac-ssa` from `fac.wast`, I noticed that the stack continues to grow. It appears that the interpreter is not removing the initial loop parameters when unwinding. We could validate that the stack is empty when the function returns.

### PR DESCRIPTION
**Pull request by andreaTP**
I ended up rewriting the entire control flow in a way that maps closely to the `TypeValidator` implementation.

Fix #401

### CODE REVIEW

**Comment by bhelx** [comment]
The SNES emulator was crashing after 100 frames or so on the interpreter. Now working on this PR!

Screenshot 2024-07-26 at 9 10 00 AM


I think this PR fixed the stack effect bug :)

**Review by bhelx — APPROVED** [review]
Looks good to me! Would be good to get a second pair of eyes on someone more familiar with the control flow algorithm.

File: runtime/src/main/java/com/dylibso/chicory/runtime/StackFrame.java [hunk 1]
```diff
@@ -89,30 +87,63 @@ public boolean terminated() {
...
 
-    public void endOfNonControlBlock() {
-        if (currentInstruction.depth() > 0) {
-            stackSizeBeforeBlock.pop();
+    public void jumpTo(int newPc) {
+        pc = newPc;
+    }
+
+    public static void doControlTransfer(CtrlFrame ctrlFrame, MStack stack) {
+        var endResults = ctrlFrame.startValues + ctrlFrame.endValues; // fix 401
```
Comment (electrum) [inline runtime/src/main/java/com/dylibso/chicory/runtime/StackFrame.java:132]:
```java
// unwind stack

**Review by electrum — APPROVED** [review]

**Review by mariofusco — APPROVED** [review]
Nice rework, much clearer now.

File: wasm-corpus/src/main/resources/wat/fac.wat
[the remark this answers is not in the document]
Comment (andreaTP) [reply wasm-corpus/src/main/resources/wat/fac.wat:1]: done

File: runtime/src/main/java/com/dylibso/chicory/runtime/StackFrame.java (hunk 1 above)
Comment (andreaTP) [reply runtime/src/main/java/com/dylibso/chicory/runtime/StackFrame.java]: changed

**Pull request merged**